search/includes/classes/class-versioning.php

```diff
@@ -745,4 +759,176 @@ public function action__ep_delete_indexable( $object_id, $indexable_slug ) {
+	}
+
+	/**
+	 * Check if the versions are persisted correctly. If not recreate them.
+	 */
+	public function maybe_self_heal() {
+		if ( $this->is_self_heal_ongoing() ) {
+			return;
+		}
+		$this->mark_self_heal_ongoing();
+
+		if ( $this->is_versioning_valid() ) {
+			$this->mark_self_heal_finished();
+			return;
+		}
+
+		$indicies = $this->get_all_accesible_indicies();
+		if ( is_wp_error( $indicies ) ) {
+			$this->mark_self_heal_finished();
+			return;
+		}
+		$versioning = $this->reconstruct_versioning_option( $indicies );
+
+		update_option( self::INDEX_VERSIONS_OPTION, $versioning );
+
+		$this->mark_self_heal_finished();
+	}
+
+	public function is_versioning_valid() {
+		$all_versions = get_option( self::INDEX_VERSIONS_OPTION, array() );
+
+		if ( ! is_array( $all_versions ) || count( $all_versions ) == 0 ) {
+			return false;
+		}
+
+		foreach ( $all_versions as $versions ) {
+			if ( ! is_array( $versions ) || count( $versions ) == 0 ) {
+				return false;
+			}
+
+			$active_statuses = wp_list_pluck( $versions, 'active' );
+
+			$array_index = array_search( true, $active_statuses, true );
+
+			if ( false === $array_index ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	public function get_all_accesible_indicies() {
+		$response = $this->elastic_search_instance->remote_request( '_cat/indices?format=json' );
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$response_code = (int) wp_remote_retrieve_response_code( $response );
+
+		if ( $response_code >= 400 ) {
+			return new \WP_Error(
+				'failed-to-fetch-indicies',
+				sprintf( 'Request failed to fetch indicies with status %s', $response_code )
+			);
+		}
+
+		$response_body_json = wp_remote_retrieve_body( $response );
+		$response_body = json_decode( $response_body_json, true );
+		$found_indices = [];
+
+		if ( ! is_array( $response_body ) ) {
+			return $found_indices;
+		}
+
+		foreach ( $response_body as $index_obj ) {
+			if ( is_array( $index_obj ) && isset( $index_obj['index'] ) ) {
+				$found_indices[] = $index_obj['index'];
+			}
+		}
+
+		return $found_indices;
+	}
+
+	public function reconstruct_versioning_option( $indicies ) {
+		if ( ! is_array( $indicies ) ) {
+			return [];
+		}
+
+		$versioning = [];
+
+		foreach ( $indicies as $index ) {
+			$index_info = $this->parse_index_name( $index );
+
+			if ( is_wp_error( $index_info ) ) {
+				continue;
+			}
+			$blog_id = get_current_blog_id();
+			if ( ! isset( $index_info['blog_id'] ) || $blog_id !== $index_info['blog_id'] ) {
+				continue;
+			}
+
+			$slug = $index_info['slug'];
+
+			if ( ! isset( $versioning[ $slug ] ) ) {
+				$versioning[ $slug ] = [];
+			}
+			$versioning[ $slug ][] = $index_info['version'];
+		}
+
+		foreach ( $versioning as $slug => $versions ) {
+			sort( $versions );
+			$version_objects = array_map( function( $version ) {
+				return [
+					'number' => $version,
+					'active' => false,
+				];
```

Comment:
Can we reuse existing functionality here somehow? This format doesn't match the existing format (missing some fields), so it would be good to keep things DRY and consistent.

Reply:
There is that normalize_version method that runs when getting versions. But as it runs on getting versions I assumed it is fine to keep. I can add it here as well though. Do you think that would be good?

Reply:
👍 that sounds good and should work for this to keep it consistent.

Reply:
added